src/blas/impl/KokkosBlas3_gemm_impl.hpp

```diff
@@ -0,0 +1,506 @@
+struct impl_gemm_choose_copy_layout<Kokkos::Cuda,LayoutA,LayoutAScratch> {
+  typedef LayoutA type;
+};
+#endif
+
+// DeepCopy matrix block into scratch
+template<class TeamHandle, class ViewTypeScratch, class ViewType, class Layout, int blockDim_i, int blockDim_j, int Transpose>
+struct impl_deep_copy_matrix_block;
+
+template<class TeamHandle, class ViewTypeScratch, class ViewType, class Layout, int blockDim_i, int blockDim_j>
+struct impl_deep_copy_matrix_block<TeamHandle,ViewTypeScratch,ViewType,Layout,blockDim_i,blockDim_j,0> {
+  typedef typename ViewType::non_const_value_type value_type;
+  typedef Kokkos::Details::ArithTraits<value_type>     ATV;
+
+  KOKKOS_INLINE_FUNCTION
+  static void copy(const TeamHandle& team, const ViewTypeScratch& A_scr, const ViewType& A, const int& offset_i, const int& offset_j) {
+    if(offset_i + blockDim_i <= A.extent(0) && offset_j + blockDim_j <= A.extent(1)) {
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockDim_j), [&] (const int j) {
+        const int idx_j = offset_j+j;
+        Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_i), [&] (const int i) {
+          const int idx_i = offset_i+i;
+          A_scr(i,j) = A(idx_i,idx_j);
+        });
+      });
+    } else {
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockDim_j), [&] (const int j) {
+        const int idx_j = offset_j+j;
+        Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_i), [&] (const int i) {
+          const int idx_i = offset_i+i;
+          A_scr(i,j) = idx_i<A.extent(0) && idx_j<A.extent(1) ? A(idx_i,idx_j) : ATV::zero();
+        });
+      });
+    }
+  }
+};
+
+template<class TeamHandle, class ViewTypeScratch, class ViewType,  int blockDim_i, int blockDim_j>
+struct impl_deep_copy_matrix_block<TeamHandle,ViewTypeScratch,ViewType,Kokkos::LayoutRight,blockDim_i,blockDim_j,0> {
+  typedef typename ViewType::non_const_value_type value_type;
+  typedef Kokkos::Details::ArithTraits<value_type>     ATV;
+
+  KOKKOS_INLINE_FUNCTION
+  static void copy(const TeamHandle& team, const ViewTypeScratch& A_scr, const ViewType& A, const int& offset_i, const int& offset_j) {
+    if(offset_i + blockDim_i <= A.extent(0) && offset_j + blockDim_j <= A.extent(1)) {
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockDim_i), [&] (const int i) {
+        const int idx_i = offset_i+i;
+        Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_j), [&] (const int j) {
+          const int idx_j = offset_j+j;
+          A_scr(i,j) = A(idx_i,idx_j);
+        });
+      });
+    } else {
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockDim_i), [&] (const int i) {
+        const int idx_i = offset_i+i;
+        Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_j), [&] (const int j) {
+          const int idx_j = offset_j+j;
+          A_scr(i,j) = idx_i<A.extent(0) && idx_j<A.extent(1) ? A(idx_i,idx_j) : ATV::zero();
+        });
+      });
+    }
+  }
+};
+
```

Comment:
A question. Do we also consider that this gemm function is also used for sacado type ? As it uses vector range here, there is no chance that AD type can use vectorization inside. So, I guess that this whole thing is only about built-in data type (with AD type, packing does not make sense and data is not even contiguous).

Can we make this clear in some documentation ?

Reply:
There is many other places where this doesn't work for those types, since we are using ThreadVectorRange in a number of math kernels (for example SPMV). This is a general concern. If you use a degree of parallelism from Kokkos internally to your data type you need to potentially specialize all KokkosKernels functions.

Reply:
@kyungjoo-kim See my comment here: https://github.com/kokkos/kokkos-kernels/issues/125#issuecomment-348351565

Stokhos, etc. has its own work-arounds for Belos adapters.  Tpetra still needs to compile correctly, though it can just wrap the GEMM implementation with a wrapper than can be partially specialized not to call those functions.